File: referencia/flutter-app/lib/ui/pantallas/apertura.dart
/// Apertura de mazorcas (RF-APE-01 a RF-APE-03).
library;

import 'package:flutter/material.dart';

import '../../servicios.dart';
import '../comun/widgets.dart';
import '../tema.dart';

class PantallaApertura extends StatefulWidget {
  const PantallaApertura({super.key, required this.loteId});
  final String loteId;

  @override
  State<PantallaApertura> createState() => _PantallaAperturaState();
}

class _PantallaAperturaState extends State<PantallaApertura> {
  final _formulario = GlobalKey<FormState>();
  final _mazorcas = TextEditingController();
  final _baba = TextEditingController();
  final _cascara = TextEditingController(text: '0');
  final _notas = TextEditingController();

  double _cascaraRecomendada = 0;
  bool _guardando = false;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) => _precargar());
  }

  Future<void> _precargar() async {
    final s = ProveedorServicios.de(context);
    final c = await s.lotes.completo(widget.loteId);
    if (!mounted) return;
    if (c?.apertura != null) {
      final a = c!.apertura!;
      setState(() {
        _mazorcas.text = '${a.mazorcasAbiertas}';
        _baba.text = '${a.kgBaba}';
        _cascara.text = '${a.kgCascaraAnadida}';
        _notas.text = a.notas;
      });
      await _recalcular();
    } else if (c?.recepcion != null) {
      setState(() {
        _mazorcas.text =
            '${c!.recepcion!.mazorcasTotal - c.recepcion!.mazorcasDescartadas}';
      });
    }
  }

  Future<void> _recalcular() async {
    final s = ProveedorServicios.de(context);
    final baba = leerNumero(_baba) ?? 0;
    final recomendada = await s.lotes.cascaraRecomendada(baba);
    if (mounted) setState(() => _cascaraRecomendada = recomendada);
  }

  @override
  void dispose() {
    for (final c in [_mazorcas, _baba, _cascara, _notas]) {
      c.dispose();
    }
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final mazorcas = leerNumero(_mazorcas) ?? 0;
    final baba = leerNumero(_baba) ?? 0;
    final porMazorca = mazorcas > 0 ? baba / mazorcas : null;

    return Scaffold(
      appBar: AppBar(title: const Text('Apertura de mazorcas')),
      body: Form(
        key: _formulario,
        child: ListView(
          padding: const EdgeInsets.all(16),
          children: [
            TarjetaSeccion(
              titulo: 'Lo que se abrió',
              icono: Icons.egg_outlined,
              hijos: [
                CampoNumero(
                  etiqueta: 'Mazorcas abiertas',
                  controlador: _mazorcas,
                  decimales: false,
                  obligatorio: true,
                  minimo: 1,
                  onChanged: (_) => setState(() {}),
                ),
                CampoNumero(
                  etiqueta: 'Baba obtenida',
                  controlador: _baba,
                  unidad: 'kg',
                  obligatorio: true,
                  minimo: 0,
                  ayuda: 'El grano fresco con pulpa, recién sacado',
                  onChanged: (_) {
                    setState(() {});
                    _recalcular();
                  },
                ),
                if (porMazorca != null)
                  FilaDato(
                    'Rendimiento por mazorca',
                    '${porMazorca.toStringAsFixed(3)} kg',
                    color: (porMazorca - 0.17).abs() / 0.17 > 0.2
                        ? ColoresEstado.atencion
                        : ColoresEstado.bien,
                  ),
                if (porMazorca != null)
                  const Text(
                    'Lo normal en CCN-51 son unos 0,170 kg por mazorca.',
                    style: TextStyle(fontSize: 14),
                  ),
              ],
            ),
            TarjetaSeccion(
              titulo: 'Cáscara añadida',
              icono: Icons.add_circle_outline,
              hijos: [
                const Text(
                  'Si hay poca baba, el montón no conserva el calor y la '
                  'fermentación se enfría. Añadir cáscara de mazorca troceada '
                  'aumenta la masa y el azúcar disponible.',
                  style: TextStyle(fontSize: 15),
                ),
                const SizedBox(height: 12),
                if (_cascaraRecomendada > 0)
                  Aviso(
                    icono: Icons.lightbulb_outline,
                    titulo: 'Te recomiendo añadir cáscara',
                    texto:
                        'Con ${baba.toStringAsFixed(1)} kg de baba te faltan '
                        'unos ${_cascaraRecomendada.toStringAsFixed(1)} kg '
                        'para llegar a la masa mínima.',
                    accion: OutlinedButton(
                      onPressed: () {
                        _cascara.text =
                            _cascaraRecomendada.toStringAsFixed(1);
                        setState(() {});
                      },
                      child: const Text('Usar esa cantidad'),
                    ),
                  ),
                CampoNumero(
                  etiqueta: 'Cáscara añadida',
                  controlador: _cascara,
                  unidad: 'kg',
                  minimo: 0,
                  onChanged: (_) => setState(() {}),
                ),
                FilaDato(
                  'Masa total al fermentador',
                  '${(baba + (leerNumero(_cascara) ?? 0)).toStringAsFixed(1)} kg',
                ),
              ],
            ),
            TarjetaSeccion(
              titulo: 'Notas',
              icono: Icons.edit_note,
              hijos: [
                TextFormField(
                  controller: _notas,
                  maxLines: 3,
                  decoration: const InputDecoration(
                    hintText: 'Lo que quieras recordar de esta apertura',
                  ),
                ),
              ],
            ),
            const SizedBox(height: 16),
            BotonGrande(
              texto: _guardando ? 'Guardando…' : 'Guardar la apertura',
              icono: Icons.save_outlined,
              onPressed: _guardando ? null : _guardar,
            ),
          ],
        ),
      ),
    );
  }

  Future<void> _guardar() async {
    if (!_formulario.currentState!.validate()) return;
    final s = ProveedorServicios.de(context);
    setState(() => _guardando = true);

    try {
      final alertas = await s.lotes.guardarApertura(
        loteId: widget.loteId,
        mazorcasAbiertas: leerNumero(_mazorcas)!.round(),
        kgBaba: leerNumero(_baba)!,
        kgCascaraAnadida: leerNumero(_cascara) ?? 0,
        notas: _notas.text.trim(),
      );
      if (!mounted) return;
      avisar(
        context,
        alertas.isEmpty
            ? 'Apertura guardada'
            : 'Apertura guardada, con ${alertas.length} aviso(s)',
      );
      Navigator.pop(context);
    } catch (e) {
      if (mounted) avisar(context, '$e', error: true);
    } finally {
      if (mounted) setState(() => _guardando = false);
    }
  }
}

File: android/app/src/main/kotlin/ec/cacaotrace/ui/pantallas/Apertura.kt
package ec.cacaotrace.ui.pantallas

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AddCircleOutline
import androidx.compose.material.icons.filled.Egg
import androidx.compose.material.icons.filled.EditNote
import androidx.compose.material.icons.filled.Save
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavHostController
import ec.cacaotrace.ContenedorApp
import ec.cacaotrace.ui.ColoresEstado
import ec.cacaotrace.ui.comun.Aviso
import ec.cacaotrace.ui.comun.BarraSuperior
import ec.cacaotrace.ui.comun.BotonGrande
import ec.cacaotrace.ui.comun.CampoNumero
import ec.cacaotrace.ui.comun.FilaDato
import ec.cacaotrace.ui.comun.Formato
import ec.cacaotrace.ui.comun.TarjetaSeccion
import ec.cacaotrace.ui.comun.leerNumero
import kotlin.math.abs
import kotlinx.coroutines.launch

/** Apertura de mazorcas (RF-APE-01 a RF-APE-03). */
@Composable
fun PantallaApertura(
    contenedor: ContenedorApp,
    navegacion: NavHostController,
    loteId: String,
) {
    val alcance = rememberCoroutineScope()
    var mazorcas by rememberSaveable { mutableStateOf("") }
    var baba by rememberSaveable { mutableStateOf("") }
    var cascara by rememberSaveable { mutableStateOf("0") }
    var notas by rememberSaveable { mutableStateOf("") }
    var cascaraRecomendada by remember { mutableStateOf(0.0) }
    var aviso by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(loteId) {
        val c = contenedor.lotes.completo(loteId)
        c?.apertura?.let { a ->
            mazorcas = a.mazorcasAbiertas.toString()
            baba = a.kgBaba.toString()
            cascara = a.kgCascaraAnadida.toString()
            notas = a.notas
        } ?: c?.recepcion?.let { r ->
            mazorcas = (r.mazorcasTotal - r.mazorcasDescartadas).toString()
        }
    }

    LaunchedEffect(baba) {
        cascaraRecomendada = contenedor.lotes.cascaraRecomendada(leerNumero(baba) ?: 0.0)
    }

    val nMazorcas = leerNumero(mazorcas) ?: 0.0
    val kgBaba = leerNumero(baba) ?: 0.0
    val porMazorca = if (nMazorcas > 0) kgBaba / nMazorcas else null

    Scaffold(topBar = { BarraSuperior("Apertura de mazorcas", navegacion) }) { relleno ->
        LazyColumn(
            Modifier.fillMaxSize().padding(relleno),
            contentPadding = PaddingValues(16.dp),
        ) {
            item {
                aviso?.let { Aviso(texto = it, color = ColoresEstado.problema) }

                TarjetaSeccion("Lo que se abrió", icono = Icons.Default.Egg) {
                    Column {
                        CampoNumero(
                            "Mazorcas abiertas", mazorcas, { mazorcas = it },
                            decimales = false,
                        )
                        CampoNumero(
                            "Baba obtenida", baba, { baba = it }, unidad = "kg",
                            ayuda = "El grano fresco con pulpa, recién sacado",
                        )
                        porMazorca?.let { r ->
                            FilaDato(
                                "Rendimiento por mazorca",
                                "${Formato.numero(r, 3)} kg",
                                color = if (abs(r - 0.17) / 0.17 > 0.2) {
                                    ColoresEstado.atencion
                                } else {
                                    ColoresEstado.bien
                                },
                            )
                            Text(
                                "Lo normal en CCN-51 son unos 0.170 kg por mazorca.",
                                fontSize = 14.sp,
                            )
                        }
                    }
                }

                TarjetaSeccion("Cáscara añadida", icono = Icons.Default.AddCircleOutline) {
                    Column {
                        Text(
                            "Si hay poca baba, el montón no conserva el calor y la " +
                                "fermentación se enfría. Añadir cáscara de mazorca " +
                                "troceada aumenta la masa y el azúcar disponible.",
                            fontSize = 15.sp,
                        )
                        if (cascaraRecomendada > 0) {
                            Aviso(
                                titulo = "Te recomiendo añadir cáscara",
                                texto = "Con ${Formato.kg(kgBaba)} de baba te faltan unos " +
                                    "${Formato.kg(cascaraRecomendada)} para llegar a la " +
                                    "masa mínima.",
                                accion = {
                                    OutlinedButton(
                                        onClick = {
                                            cascara = Formato.numero(cascaraRecomendada)
                                        },
                                    ) { Text("Usar esa cantidad") }
                                },
                            )
                        }
                        CampoNumero(
                            "Cáscara añadida", cascara, { cascara = it }, unidad = "kg",
                        )
                        FilaDato(
                            "Masa total al fermentador",
                            Formato.kg(kgBaba + (leerNumero(cascara) ?: 0.0)),
                        )
                    }
                }

                TarjetaSeccion("Notas", icono = Icons.Default.EditNote) {
                    OutlinedTextField(
                        value = notas,
                        onValueChange = { notas = it },
                        placeholder = { Text("Lo que quieras recordar de esta apertura") },
                        minLines = 3,
                        modifier = Modifier.fillMaxWidth(),
                    )
                }

                Spacer(Modifier.height(16.dp))
                BotonGrande("Guardar la apertura", Icons.Default.Save) {
                    if (nMazorcas <= 0 || kgBaba <= 0) {
                        aviso = "Escribe las mazorcas abiertas y los kg de baba"
                        return@BotonGrande
                    }
                    alcance.launch {
                        contenedor.lotes.guardarApertura(
                            loteId = loteId,
                            mazorcasAbiertas = nMazorcas.toInt(),
                            kgBaba = kgBaba,
                            kgCascaraAnadida = leerNumero(cascara) ?: 0.0,
                            notas = notas.trim(),
                        )
                        navegacion.popBackStack()
                    }
                }
            }
        }
    }
}
